### ISSUE TITLE
Button with role="switch" pronounced as checkbox.

### ISSUE THREAD

**Issue by erikkroes** [nvaccess/nvda#11310]
When navigating to a `<button>` with `role="switch"`, it will be pronounced as a checkbox. It is recognized as a toggle button or switch button by accessibility APIs for example, but it is not pronounced as such. 
https://www.w3.org/TR/wai-aria-1.2/#switch does state "A type of checkbox that represents on/off values", but it is not a checkbox and has a seperate role. I think it should be pronounced differently.
https://www.w3.org/TR/wai-aria-1.1/#switch does mention "Role: ROLE_SYSTEM_CHECKBUTTON" which might be the cause, but it also states "Object Attribute: xml-roles:switch" which hopefully means thiscan be fixed.

### Steps to reproduce:
1. Open https://codepen.io/erikkroes/pen/WNrOdWp in Chrome version 83.0.4103.116
2. Navigate to the example `<button>` with `role="switch"`

### Actual behavior:
Button is announced as "Checkbox".

### Expected behavior:
Button is announced as "Switch", or something alike.

### System configuration
NVDA 2020.1 installed on Parallels. Chrome version 83. Windows 10.

**Comment by Adriani90** [comment]
cc: @feerrenrut, @leonardder is there a way to fix this in NVDA? Actually this element can be navigated by pressing b in browse mode, so it could cause confusions if it is still not pronounced correctly.

**Comment by JulienCochuyt** [comment]
Switches are visually buttons, and functionally checkboxes...
Thus, they can currently - and IMHO should remain - reachable by both quick nav by buttons and quick nav by checkboxes.
Sure, announcing them as "Switch buttons" could improve communication with a sighted assistance 

What I feel is much more problematic in the current implementation - and demonstrated by the provided example - is that NVDA does not announce the value embedded in the Switch Button. This value can carry important information. Narrator.does announce it.

**Comment by JulienCochuyt** [comment]
The mapping of Switch Button to the checkbox role has been established for IA2 as of issue #9187 / PR #10782.
cc @jcsteh

According to https://github.com/nvaccess/nvda/issues/9187#issuecomment-458672152, proper support for `role="switch"` is still missing from IE11. I have a big IE11 WebModule in the pipe due before the end of this summer, thus I'll probably have to fix this lack pretty soon.

A uniform stance, reducing the behavioral gap between the different supported browsers would IMHO be much preferable.

**Comment by Brian1Gaff** [comment]
For those who don't know, Windows 7 users had Microsoft force an update last 
week to the  release version of Edge with text to the effect of that IE11 
was now considered obsolete and a security issue. I did not read it fully, I 
have to say, but many web sites now put up a banner telling people who use 
it to use something else, and some of  the in page audio players are 
beginning to not support ie at all.

They did not remove it, but it is looking to me like there is a push by 
Microsoft to  stop people using the old IE generally.
 Brian

bglists@blueyonder.co.uk
Sent via blueyonder.
Please address personal E-mail to:-
briang1@blueyonder.co.uk, putting 'Brian Gaff'
in the display name field.
Newsgroup monitored: alt.comp.blind-users
----- Original Message -----

**Comment by JulienCochuyt** [comment]
@Brian1Gaff, surely IE11 has reached end of life. Thanks for this warning.
Still, as long as I will be enrolled to circumvent accessibility/ergonomics issues on business web applications old enough to forcibly require it, I'll do my best to ensure NVDA supports it as much as reason dictates for the sake of our fellow users.

@feerrenrut, could you please mark this very comment and Brian's as off-topic for better readability of this issue? (or empower me do to so)

**Comment by jcsteh** [comment]
> https://www.w3.org/TR/wai-aria-1.2/#switch does state "A type of checkbox that represents on/off values", but it is not a checkbox and has a seperate role. I think it should be pronounced differently.

IMO, toggle button isn't any better than check box. If anything, it's worse because buttons normally have pressed/not pressed states, rather than checked/not checked.

You note that the spec says "a type of checkbox", but you then say "it is not a checkbox". If it's a type of checkbox, it is by definition a checkbox, albeit slightly different to a vanilla checkbox.

Then there's the question of whether it's useful (or even harmful) for users to understand the difference between a checkbox and a switch, especially on desktop. Desktop screen reader users know what a checkbox is. Do they know what a switch is and how to interact with it? Does understanding this difference help users or just give them more things to learn, making the already steep learning curve that tiny bit steeper?

Just my thoughts; this really isn't my decision.

In terms of implementation, if it was decided to change this, we'd need a new NVDA ROLE_SWITCH. We might also need new states for on/off, or alternatively, speech/braille would somehow need to know to present checked/not checked as on/off for a switch. Finally, NVDAObjects.IAccessible.ia2Web.Switch would need to be updated accordingly.

**Comment by leonardder** [comment]
> What I feel is much more problematic in the current implementation - and demonstrated by the provided example - is that NVDA does not announce the value embedded in the Switch Button. This value can carry important information. Narrator.does announce it.

This applies to all check boxes and was a show stopper for #8858 to use the values for the remaining columns after the first columns The problem is, how to guarantee that the value has meaningful information? Some implementation expose the checked/unchecked state as 0/1 in the accValue, and that's really redundant information.

May be this is worth a separate issue?

**Comment by jcsteh** [comment]
I don't know what is meant by "value" here. An ARIA checkbox can't specify a value. Narrator says on and off for switches instead of checked/not checked, but that's just how Narrator pronounces switches.

**Comment by JulienCochuyt** [comment]
> Does understanding this difference help users or just give them more things to learn

The confusion may interfere when collaborating with a sighted party.

> An ARIA checkbox can't specify a value. 

Indeed, but a switch can, and its value is currently unavailable to NVDA users.

> Some implementation expose the checked/unchecked state as 0/1 in the accValue, and that's really redundant information.

Do you know which implementations? Could we black- or white- list this? Or filter out obvious redundancies?

**Comment by jcsteh** [comment]
> An ARIA switch cannot specify a value according to the spec:
https://w3c.github.io/aria/#switch
Note that aria-valuetext is not listed as a supported attribute. Note also that nothing maps to accValue in the mapping spec for switch:
https://w3c.github.io/core-aam/#mapping_role_table

**Comment by jcsteh** [comment]
If you're referring to the text content inside the tag, that is not a value. In fact, the spec says that children of a switch are presentational; they relate only to how the control is presented visually.

**Comment by erikkroes** [comment]
Any update on this? What's the plan?

**Comment by feerrenrut** [comment]
It looks this issue is stuck. There is some debate about whether this change should be made, though it does look like it could be made.

There is also the discussion on W3C about how switches should be exposed: https://github.com/w3c/core-aam/issues/75

I think this probably should be exposed to the user, since it encourages using a common language when talking about the control, and differentiates a two state switch from a three state (checked, not checked, indeterminate) checkbox

**Comment by kpasagada** [comment]
Is there an update on this issue? It looks like its been stuck for a while now.

**Comment by feerrenrut** [comment]
On the whole I think this should go ahead, summary of reasoning giving in this thread (for and against):
- It should be a small low-risk change
- It isn't technically required, checked/not checked conveys the Boolean nature of switches and toggle buttons.
- Allows easier communication between SR and non-SR users.
- Prevent hacky approximations via role-description
- Give web authors more control over the speech experience.
- SR users need to learn and familiarize themselves with more controls.

The note on https://www.w3.org/TR/wai-aria-1.2/#switch also reflects this.
>NOTE
>A switch provides approximately the same functionality as a checkbox and toggle button, but makes it possible for assistive technologies to present the widget in a fashion consistent with its on-screen appearance.

https://github.com/nvaccess/nvda/issues/10644 is also related.

**Comment by feerrenrut** [comment]
Please see closed duplicates:
- https://github.com/nvaccess/nvda/issues/12306
- https://github.com/nvaccess/nvda/issues/13647

**Comment by seanbudd** [comment]
Blocked by https://github.com/w3c/core-aam/issues/75
